Index: src/freeglut_main.c
===================================================================
--- src/freeglut_main.c	(revision 915)
+++ src/freeglut_main.c	(working copy)
@@ -70,6 +70,12 @@
 #    define MIN(a,b) (((a)<(b)) ? (a) : (b))
 #endif
 
+#ifdef WM_TOUCH
+    typedef BOOL (*pGetTouchInputInfo)(HTOUCHINPUT,UINT,PTOUCHINPUT,int);
+    typedef BOOL (*pCloseTouchInputHandle)(HTOUCHINPUT);
+	static pGetTouchInputInfo fghGetTouchInputInfo = (pGetTouchInputInfo)0xDEADBEEF;
+	static pCloseTouchInputHandle fghCloseTouchInputHandle = (pCloseTouchInputHandle)0xDEADBEEF;
+#endif
 
 /*
  * TODO BEFORE THE STABLE RELEASE:
@@ -2455,7 +2461,18 @@
 		unsigned int numInputs = (unsigned int)wParam;
 		unsigned int i = 0;
 		TOUCHINPUT* ti = (TOUCHINPUT*)malloc( sizeof(TOUCHINPUT)*numInputs);
-		if (GetTouchInputInfo( (HTOUCHINPUT)lParam, numInputs, ti, sizeof(TOUCHINPUT) )) {
+
+		if (fghGetTouchInputInfo == (pGetTouchInputInfo)0xDEADBEEF) {
+		    fghGetTouchInputInfo = (pGetTouchInputInfo)GetProcAddress(GetModuleHandle("user32"),"GetTouchInputInfo");
+		    fghCloseTouchInputHandle = (pCloseTouchInputHandle)GetProcAddress(GetModuleHandle("user32"),"CloseTouchInputHandle");
+		}
+
+		if (!fghGetTouchInputInfo) { 
+			free( (void*)ti );
+			break;
+		}
+
+		if (fghGetTouchInputInfo( (HTOUCHINPUT)lParam, numInputs, ti, sizeof(TOUCHINPUT) )) {
 			/* Handle each contact point */
 			for (i = 0; i < numInputs; ++i ) {
 
@@ -2477,7 +2494,7 @@
 				}
 			}
 		}
-		CloseTouchInputHandle((HTOUCHINPUT)lParam);
+		fghCloseTouchInputHandle((HTOUCHINPUT)lParam);
 		free( (void*)ti );
 		lRet = 0; /*DefWindowProc( hWnd, uMsg, wParam, lParam );*/
 		break;
Index: src/freeglut_window.c
===================================================================
--- src/freeglut_window.c	(revision 915)
+++ src/freeglut_window.c	(working copy)
@@ -141,6 +141,10 @@
 
 #endif  /* TARGET_HOST_MS_WINDOWS */
 
+#ifdef WM_TOUCH
+	typedef BOOL (*pRegisterTouchWindow)(HWND,ULONG);
+   static pRegisterTouchWindow fghRegisterTouchWindow = (pRegisterTouchWindow)0xDEADBEEF;
+#endif
 
 /* pushing attribute/value pairs into an array */
 #define ATTRIB(a) attributes[where++]=(a)
@@ -1293,7 +1297,10 @@
 
     /* Enable multitouch: additional flag TWF_FINETOUCH, TWF_WANTPALM */
     #ifdef WM_TOUCH
-       RegisterTouchWindow( window->Window.Handle, TWF_FINETOUCH | TWF_WANTPALM );
+        if (fghRegisterTouchWindow == (pRegisterTouchWindow)0xDEADBEEF) 
+			fghRegisterTouchWindow = (pRegisterTouchWindow)GetProcAddress(GetModuleHandle("user32"),"RegisterTouchWindow");
+		if (fghRegisterTouchWindow)
+             fghRegisterTouchWindow( window->Window.Handle, TWF_FINETOUCH | TWF_WANTPALM );
     #endif
 
 #if defined(_WIN32_WCE)
