Index: src/freeglut_window.c
===================================================================
--- src/freeglut_window.c	(Revision 904)
+++ src/freeglut_window.c	(Arbeitskopie)
@@ -1120,6 +1120,11 @@
         window->Window.Context
     );
 
+    /* register extension events _before_ window is mapped */
+    #ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
+       fgRegisterDevices( fgDisplay.Display, &(window->Window.Handle) );
+    #endif
+
     XMapWindow( fgDisplay.Display, window->Window.Handle );
 
     XFree(visualInfo);
@@ -1286,6 +1291,10 @@
 /*  SetWindowPos(window->Window.Handle, NULL, 0, 0, 0, 0,
      SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER | SWP_FRAMECHANGED); */
 
+    /* Enable multitouch: additional flag TWF_FINETOUCH, TWF_WANTPALM */
+    #ifdef WM_TOUCH
+       RegisterTouchWindow( window->Window.Handle, TWF_FINETOUCH | TWF_WANTPALM );
+    #endif
 
 #if defined(_WIN32_WCE)
     ShowWindow( window->Window.Handle, SW_SHOW );
Index: src/freeglut_callbacks.c
===================================================================
--- src/freeglut_callbacks.c	(Revision 904)
+++ src/freeglut_callbacks.c	(Arbeitskopie)
@@ -373,4 +373,40 @@
     SET_CALLBACK( TabletButton );
 }
 
+/*
+ * Sets the multi-pointer entry callback for the current window
+ */
+void FGAPIENTRY glutMultiEntryFunc( void (* callback)(int, int ) )
+{
+    FREEGLUT_EXIT_IF_NOT_INITIALISED ( "glutMultiEntryFunc" );
+    SET_CALLBACK( MultiEntry );
+}
+
+/*
+ * Sets the multi-pointer button callback for the current window
+ */
+void FGAPIENTRY glutMultiButtonFunc( void (* callback)(int, int, int, int, int ) )
+{
+    FREEGLUT_EXIT_IF_NOT_INITIALISED ( "glutMultiButtonFunc" );
+    SET_CALLBACK( MultiButton );
+}
+
+/*
+ * Sets the multi-pointer motion callback for the current window
+ */
+void FGAPIENTRY glutMultiMotionFunc( void (* callback)(int, int, int ) )
+{
+    FREEGLUT_EXIT_IF_NOT_INITIALISED ( "glutMultiMotionFunc" );
+    SET_CALLBACK( MultiMotion );
+}
+
+/*
+ * Sets the multi-pointer passive motion callback for the current window
+ */
+void FGAPIENTRY glutMultiPassiveFunc( void (* callback)(int, int, int ) )
+{
+    FREEGLUT_EXIT_IF_NOT_INITIALISED ( "glutMultiPassiveFunc" );
+    SET_CALLBACK( MultiPassive );
+}
+
 /*** END OF FILE ***/
Index: src/freeglut_internal.h
===================================================================
--- src/freeglut_internal.h	(Revision 904)
+++ src/freeglut_internal.h	(Arbeitskopie)
@@ -238,6 +238,11 @@
 typedef void (* FGCBTabletButton  )( int, int, int, int );
 typedef void (* FGCBDestroy       )( void );
 
+typedef void (* FGCBMultiEntry   )( int, int );
+typedef void (* FGCBMultiButton  )( int, int, int, int, int );
+typedef void (* FGCBMultiMotion  )( int, int, int );
+typedef void (* FGCBMultiPassive )( int, int, int );
+
 /* The global callbacks type definitions */
 typedef void (* FGCBIdle          )( void );
 typedef void (* FGCBTimer         )( int );
@@ -587,6 +592,12 @@
     CB_Joystick,
     CB_Destroy,
 
+    /* MPX-related */
+    CB_MultiEntry,
+    CB_MultiButton,
+    CB_MultiMotion,
+    CB_MultiPassive,
+
     /* Presently ignored */
     CB_Select,
     CB_OverlayDisplay,
@@ -970,8 +981,15 @@
  */
 #if TARGET_HOST_POSIX_X11
 int fgHintPresent(Window window, Atom property, Atom hint);
+
+/* Handler for X extension Events */
+#ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
+  void fgHandleExtensionEvents( XEvent * ev );
+  void fgRegisterDevices( Display* dpy, Window* win );
 #endif
 
+#endif
+
 SFG_Proc fghGetProcAddress( const char *procName );
 
 #if TARGET_HOST_MS_WINDOWS
Index: src/freeglut_main.c
===================================================================
--- src/freeglut_main.c	(Revision 904)
+++ src/freeglut_main.c	(Arbeitskopie)
@@ -512,7 +512,7 @@
 /*
  * Returns GLUT modifier mask for the state field of an X11 event.
  */
-static int fghGetXModifiers( int state )
+int fghGetXModifiers( int state )
 {
     int ret = 0;
 
@@ -1445,7 +1445,10 @@
             break;
 
         default:
-            fgWarning ("Unknown X event type: %d\n", event.type);
+            /* enter handling of Extension Events here */
+            #ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
+                fgHandleExtensionEvents( &event );
+            #endif
             break;
         }
     }
@@ -2444,6 +2447,41 @@
         lRet = DefWindowProc( hWnd, uMsg, wParam, lParam );
         break;
 
+#ifdef WM_TOUCH
+	/* handle multi-touch messages */
+	case WM_TOUCH:
+	{
+		unsigned int numInputs = (unsigned int)wParam;
+		unsigned int i = 0;
+		TOUCHINPUT* ti = (TOUCHINPUT*)malloc( sizeof(TOUCHINPUT)*numInputs);
+		if (GetTouchInputInfo( (HTOUCHINPUT)lParam, numInputs, ti, sizeof(TOUCHINPUT) )) {
+			/* Handle each contact point */
+			for (i = 0; i < numInputs; ++i ) {
+
+				POINT tp;
+				tp.x = TOUCH_COORD_TO_PIXEL(ti[i].x);
+				tp.y = TOUCH_COORD_TO_PIXEL(ti[i].y);
+				ScreenToClient( hWnd, &tp );
+
+				ti[i].dwID = ti[i].dwID * 2;
+
+				if (ti[i].dwFlags & TOUCHEVENTF_DOWN) {
+					INVOKE_WCB( *window, MultiEntry,  ( ti[i].dwID, GLUT_ENTERED ) );
+					INVOKE_WCB( *window, MultiButton, ( ti[i].dwID, tp.x, tp.y, 0, GLUT_DOWN ) );
+				} else if (ti[i].dwFlags & TOUCHEVENTF_MOVE) {
+					INVOKE_WCB( *window, MultiMotion, ( ti[i].dwID, tp.x, tp.y ) );
+				} else if (ti[i].dwFlags & TOUCHEVENTF_UP)   { 
+					INVOKE_WCB( *window, MultiButton, ( ti[i].dwID, tp.x, tp.y, 0, GLUT_UP ) );
+					INVOKE_WCB( *window, MultiEntry,  ( ti[i].dwID, GLUT_LEFT ) );
+				}
+			}
+		}
+		CloseTouchInputHandle((HTOUCHINPUT)lParam);
+		free( (void*)ti );
+		lRet = 0; /*DefWindowProc( hWnd, uMsg, wParam, lParam );*/
+		break;
+	}
+#endif
     default:
         /* Handle unhandled messages */
         lRet = DefWindowProc( hWnd, uMsg, wParam, lParam );
Index: src/freeglut_xinput.c
===================================================================
--- src/freeglut_xinput.c	(Revision 0)
+++ src/freeglut_xinput.c	(Revision 0)
@@ -0,0 +1,219 @@
+/* Written for XI1 by Nikolas Doerfler <doerflen@in.tum.de> (c) 2008 *
+ * Rewritten for XI2 by Florian Echtler <echtler@in.tum.de> (c) 2009 */
+
+#include <GL/freeglut.h>
+
+#include "freeglut_internal.h"
+
+#if TARGET_HOST_POSIX_X11 && HAVE_X11_EXTENSIONS_XINPUT2_H
+
+#include <errno.h>
+#include <stdarg.h>
+
+#include <X11/Xlib.h>
+#include <X11/extensions/XInput2.h>
+
+/* import function from freeglut_main.c */
+int fghGetXModifiers( int state );
+
+/* extension opcode for XInput */
+int xi_opcode = -1;
+
+/**
+ * \brief Sets window up for XI2 events.
+ */
+void fgRegisterDevices( Display* dpy, Window* win ) {
+
+	XIEventMask mask;
+	unsigned char flags[2] = { 0, 0 };
+	int event, error;
+
+	/*Display* dpy = fgDisplay.Display;
+	Window* win = glutGetXWindow();*/
+
+	/* get XInput extension opcode */
+	if (!XQueryExtension( dpy, "XInputExtension", &xi_opcode, &event, &error )) { xi_opcode = -1; }
+
+	/* Select for motion events */
+	mask.deviceid = XIAllMasterDevices;
+	mask.mask_len = 2;
+	mask.mask = flags;
+
+	XISetMask(mask.mask, XI_Enter);
+	XISetMask(mask.mask, XI_Motion);
+	XISetMask(mask.mask, XI_ButtonPress);
+	XISetMask(mask.mask, XI_ButtonRelease);
+	XISetMask(mask.mask, XI_Leave);
+	/*XISetMask(mask.mask, XI_KeyPress);
+	XISetMask(mask.mask, XI_KeyRelease);
+	XISetMask(mask.mask, XI_DeviceChanged);
+	XISetMask(mask.mask, XI_RawEvent);
+	XISetMask(mask.mask, XI_FocusIn);
+	XISetMask(mask.mask, XI_FocusOut);
+	XISetMask(mask.mask, XI_HierarchyChanged);*/
+
+	XISelectEvents( dpy, *win, &mask, 1 );
+}
+
+
+void fgPrintXILeaveEvent(XILeaveEvent* event)
+{
+    char* mode = "";
+		char* detail = "";
+    int i;
+
+    printf("    windows: root 0x%lx event 0x%lx child 0x%ld\n",
+            event->root, event->event, event->child);
+    switch(event->mode)
+    {
+        case NotifyNormal:       mode = "NotifyNormal"; break;
+        case NotifyGrab:         mode = "NotifyGrab"; break;
+        case NotifyUngrab:       mode = "NotifyUngrab"; break;
+        case NotifyWhileGrabbed: mode = "NotifyWhileGrabbed"; break;
+    }
+    switch (event->detail)
+    {
+        case NotifyAncestor: detail = "NotifyAncestor"; break;
+        case NotifyVirtual: detail = "NotifyVirtual"; break;
+        case NotifyInferior: detail = "NotifyInferior"; break;
+        case NotifyNonlinear: detail = "NotifyNonlinear"; break;
+        case NotifyNonlinearVirtual: detail = "NotifyNonlinearVirtual"; break;
+        case NotifyPointer: detail = "NotifyPointer"; break;
+        case NotifyPointerRoot: detail = "NotifyPointerRoot"; break;
+        case NotifyDetailNone: detail = "NotifyDetailNone"; break;
+    }
+    printf("    mode: %s (detail %s)\n", mode, detail);
+    printf("    flags: %s %s\n", event->focus ? "[focus]" : "",
+                                 event->same_screen ? "[same screen]" : "");
+    printf("    buttons:");
+    for (i = 0; i < event->buttons.mask_len * 8; i++)
+        if (XIMaskIsSet(event->buttons.mask, i))
+            printf(" %d", i);
+    printf("\n");
+
+    printf("    modifiers: locked 0x%x latched 0x%x base 0x%x\n",
+            event->mods.locked, event->mods.latched,
+            event->mods.base);
+    printf("    group: locked 0x%x latched 0x%x base 0x%x\n",
+            event->group.locked, event->group.latched,
+            event->group.base);
+
+    printf("    root x/y:  %.2f / %.2f\n", event->root_x, event->root_y);
+    printf("    event x/y: %.2f / %.2f\n", event->event_x, event->event_y);
+
+}
+
+
+void fgPrintXIDeviceEvent(XIDeviceEvent* event)
+{
+    double *val;
+    int i;
+
+    printf("    device: %d (%d)\n", event->deviceid, event->sourceid);
+    printf("    detail: %d\n", event->detail);
+    printf("    buttons:");
+ 	  for (i = 0; i < event->buttons.mask_len * 8; i++)
+   	    if (XIMaskIsSet(event->buttons.mask, i))
+     	      printf(" %d", i);
+    printf("\n");
+
+    printf("    modifiers: locked 0x%x latched 0x%x base 0x%x\n",
+            event->mods.locked, event->mods.latched,
+            event->mods.base);
+    printf("    group: locked 0x%x latched 0x%x base 0x%x\n",
+            event->group.locked, event->group.latched,
+            event->group.base);
+    printf("    valuators:");
+
+    val = event->valuators.values;
+    for (i = 0; i < event->valuators.mask_len * 8; i++)
+        if (XIMaskIsSet(event->valuators.mask, i))
+            printf(" %d: %.2f", i, *val++);
+    printf("\n");
+
+    printf("    windows: root 0x%lx event 0x%lx child 0x%ld\n",
+            event->root, event->event, event->child);
+    printf("    root x/y:  %.2f / %.2f\n", event->root_x, event->root_y);
+    printf("    event x/y: %.2f / %.2f\n", event->event_x, event->event_y);
+
+}
+
+
+/**
+ * \brief This function is called when an Extension Event is received
+ * and calls the corresponding callback functions for these events.
+ */
+void fgHandleExtensionEvents( XEvent* base_ev ) {
+
+	int i, button = 0;
+	XGenericEventCookie* cookie = (XGenericEventCookie*)&(base_ev->xcookie);
+
+	if ( XGetEventData( fgDisplay.Display, cookie ) && (cookie->type == GenericEvent) && (cookie->extension == xi_opcode) ) {
+
+		XIDeviceEvent* event = (XIDeviceEvent*)(cookie->data);
+		/*printf("XI2 event type: %d - %d\n", cookie->evtype, event->type );*/
+
+		SFG_Window* window = fgWindowByHandle( event->event );
+		if (!window) return;
+
+		switch (cookie->evtype) {
+
+			case XI_Enter:
+			case XI_Leave:
+				fgState.Modifiers = fghGetXModifiers( ((XIEnterEvent*)event)->mods.base );
+				INVOKE_WCB( *window, MultiEntry, (
+					event->deviceid, 
+					(event->evtype == XI_Enter ? GLUT_ENTERED : GLUT_LEFT)
+				));
+				#if _DEBUG
+					fgPrintXILeaveEvent((XILeaveEvent*)event);
+				#endif
+				break;
+
+			case XI_ButtonPress:
+			case XI_ButtonRelease:
+				fgState.Modifiers = fghGetXModifiers( event->mods.base );
+				INVOKE_WCB( *window, MultiButton, (
+					event->deviceid, 
+					event->event_x,
+					event->event_y,
+					(event->detail)-1, 
+					(event->evtype == XI_ButtonPress ? GLUT_DOWN : GLUT_UP)
+				));
+				INVOKE_WCB( *window, Mouse, (
+					(event->detail)-1, 
+					(event->evtype == XI_ButtonPress ? GLUT_DOWN : GLUT_UP),
+					event->event_x,
+					event->event_y
+				));
+				break;
+
+			case XI_Motion:
+				fgState.Modifiers = fghGetXModifiers( event->mods.base );
+				for (i = 0; i < event->buttons.mask_len; i++) if (event->buttons.mask[i]) button = 1;
+				if (button) {
+					INVOKE_WCB( *window, MultiMotion,  ( event->deviceid, event->event_x, event->event_y ) );
+					INVOKE_WCB( *window,           Motion,  (                  event->event_x, event->event_y ) );
+				} else {
+					INVOKE_WCB( *window, MultiPassive, ( event->deviceid, event->event_x, event->event_y ) );
+					INVOKE_WCB( *window,           Passive, (                  event->event_x, event->event_y ) );
+				}
+				#if _DEBUG
+					fgPrintXIDeviceEvent(event);
+				#endif
+				break;
+
+			default:
+				#if _DEBUG
+					fgWarning( "Unknown XI2 device event:" );
+					fgPrintXIDeviceEvent( event );
+				#endif
+				break;
+		}
+		fgState.Modifiers = INVALID_MODIFIERS;
+	}
+	XFreeEventData( fgDisplay.Display, cookie );
+}
+
+#endif
+
Index: src/Makefile.am
===================================================================
--- src/Makefile.am	(Revision 904)
+++ src/Makefile.am	(Arbeitskopie)
@@ -38,7 +38,8 @@
 			 freeglut_structure.c \
 			 freeglut_teapot.c \
 			 freeglut_videoresize.c \
-			 freeglut_window.c
+			 freeglut_window.c \
+			 freeglut_xinput.c
 
 #
 # Additional linker flags
Index: src/freeglut_ext.c
===================================================================
--- src/freeglut_ext.c	(Revision 904)
+++ src/freeglut_ext.c	(Arbeitskopie)
@@ -85,6 +85,10 @@
     CHECK_NAME(glutReshapeFunc);
     CHECK_NAME(glutKeyboardFunc);
     CHECK_NAME(glutMouseFunc);
+    CHECK_NAME(glutMultiEntryFunc);
+    CHECK_NAME(glutMultiMotionFunc);
+    CHECK_NAME(glutMultiButtonFunc);
+    CHECK_NAME(glutMultiPassiveFunc);
     CHECK_NAME(glutMotionFunc);
     CHECK_NAME(glutPassiveMotionFunc);
     CHECK_NAME(glutEntryFunc);
Index: configure.ac
===================================================================
--- configure.ac	(Revision 904)
+++ configure.ac	(Arbeitskopie)
@@ -50,6 +51,7 @@
 AC_HEADER_TIME
 AC_CHECK_HEADERS([X11/extensions/xf86vmode.h], [], [], [#include <X11/Xlib.h>])
 AC_CHECK_HEADERS([X11/extensions/XI.h X11/extensions/XInput.h])
+AC_CHECK_HEADERS([X11/extensions/XInput2.h])
 CPPFLAGS="$save_CPPFLAGS"
 
 # Checks for library functions.
Index: include/GL/freeglut_ext.h
===================================================================
--- include/GL/freeglut_ext.h	(Revision 904)
+++ include/GL/freeglut_ext.h	(Arbeitskopie)
@@ -173,6 +173,17 @@
 FGAPI GLUTproc FGAPIENTRY glutGetProcAddress( const char *procName );
 
 /*
+ * Multi-touch/multi-pointer extensions
+ */
+
+#define GLUT_HAS_MULTI 1
+
+FGAPI void FGAPIENTRY glutMultiEntryFunc( void (* callback)( int, int ) );
+FGAPI void FGAPIENTRY glutMultiButtonFunc( void (* callback)( int, int, int, int, int ) );
+FGAPI void FGAPIENTRY glutMultiMotionFunc( void (* callback)( int, int, int ) );
+FGAPI void FGAPIENTRY glutMultiPassiveFunc( void (* callback)( int, int, int ) );
+
+/*
  * Joystick functions, see freeglut_joystick.c
  */
 /* USE OF THESE FUNCTIONS IS DEPRECATED !!!!! */
